tensorboard/data/provider.py

```diff
@@ -73,6 +73,22 @@ class DataProvider(object):
     for 3 random inputs from this batch". A single blob can of course be
     represented as a blob sequence that always has exactly one element.
 
+    When reading time series, *downsampling* refers to selecting a
+    subset of the points in each time series. Downsampling only occurs
+    across the step axis (rather than, e.g., the blobs in a single blob
+    sequence datum), and occurs individually within each time series.
+    When downsampling, the latest datum should always be included in the
+    sample, so that clients have a view of metrics that is maximally up
+    to date. Implementations may choose to force the first (oldest)
+    datum to be included in each sample as well, but this is not
+    required. The remainder of the points in the sample should be
```

Comment:
Re the oldest datum: if this is useful, we should require it.  If it's not, why mention it at all?  The whole point of a contract is that the caller need know nothing of implementation details.  Also, if that detail is a side effect, so be it, but implementations should not go to extra lengths to add features that are not in the spec.  As we've already seen, that is a recipe for confusion.

Reply:
For clients to be blind to implementation details, we must clearly
define which parts of the interface *are* implementation details and
which are specified. Implementation details generally extend to
“everything not explicitly mandated by the spec”, but sometimes it can
be helpful to call these out explicitly. This is well established
practice (e.g., `java.util.Map` re: null entries; `java.util.Comparable`
re: compatibility with `equals`; Python <3.7 re: dictionary order;
Bazel re: `query` output sorting; basically all the POSIX utility specs;
C/C++ re: struct layout; all the non-normative notes in ECMAScript and C
and C++).

The concrete reason that I mentioned this is that some known data
provider implementations include “newest and `k - 1` uniform” while
others include “newest, oldest, and `k - 2` uniform”. Both of these are
fine. If we omit this sentence (“…may choose…”), then the latter kind of
implementation becomes invalid, because the points other than the newest
are clearly not selected uniformly at random.

If the implication is that the contract must specify every observable
behavior of the interface, then I must disagree. Every such constraint
by definition precludes some implementations, and often includes entire
classes of implementations with desirable properties. See above list for
lots of precedent.